Edit resampling with code2 (#44)
resampling-stats/resampling-with: source/resampling_with_code2.Rmd

# More resampling with code

@sec-resampling-code introduced a problem in probability, that was also a problem in statistics.  We asked how surprised we should be at the results of a trial of a new cancer treatment regime.

> Here we study another urgent problem in the real world—racial bias and the

Reviewer: I think the original was right ...  Do you mean "real-world racial bias"?  

```suggestion
Here we study another urgent problem in the real world - racial bias and the
```

Author: Ah, yes, I am used to https://www.thepunctuationguide.com/em-dash.html#:~:text=The%20em%20dash%20is%20typically,and%20after%20the%20em%20dash.

the eligible jurors in that county were Black, but every member of Swain's
jury was White.  Swain and his legal team appealed to the Alabama Supreme
Court, and then to the [US Supreme
Court](https://en.wikipedia.org/wiki/Swain_v._Alabama), arguing

> that Black jurors were very rarely selected for juries in Talladegna county, even though they made up about a quarter of the eligible pool of jurors.

Reviewer: Original spelling correct I believe:

https://en.wikipedia.org/wiki/Talladega_County,_Alabama

```suggestion
that Black jurors were very rarely selected for juries in Talladega county, even though they made up about a quarter of the eligible pool of jurors.
```

Author: Typo, sorry!

The US Supreme Court rejected this argument, in a 6 to 3 opinion, writing that
"The overall percentage disparity has been small and reflects no studied
attempt to include or exclude a specified number of Negros.".

> And while the courts heard a lot of evidence on bias in jury selection, one observation made by Swain and his team stands out: that Swain's own jury was *entirely* White.

Reviewer: Hmm - but I don't know whether that's true - that the observation "stands out".  It was I think less surprising than a lot of the other evidence they put forward.

Author: I meant that it stands out now: that it is the type of thing that makes you raise an eyebrow.  Suggestions on how to word that?

Reviewer: How about:

```suggestion
Swain's team presented a lot of evidence on bias in jury selection, but here we will look at the obvious and apparently surprising fact that Swain's jury was entirely White.
```

The Supreme Court decided that the "disparity" between selection of White and Black jurors "has been small" — but how would they, and how would we, make a rational decision about whether this disparity *really* was "small"?

## A small disparity and a hypothetical world

To answer the question that the Supreme Court asked, we return to the method we used in the last chapter.

Let us imagine a hypothetical world, in which each individual Black or White
person had an equal chance of being selected for the jury.  Call this world Hypothetical County, Alabama.

Just as in 1960's Talladega County, 26% of eligible jurors in Hypothetical
County are Black.  Hypothetical County jury selection has no bias against

> Black men, so we expect around 26% of the jury to be Black. 0.26 * 12 =

Reviewer: The jurors were indeed restricted to men, at the time, but maybe we can miss that detail out:

```suggestion
Black people, so we expect around 26% of the jury to be Black. 0.26 * 12 =
```

Author: Right, the 60s! I completely forgot. Happy to change this back to be factually accurate if you want.

evidence that the disparity in selection between Black and White jurors, was
not "small".

::: {.question}
What is the probability of an *all White* jury being randomly selected out of a population having 26% Black people?
:::

## Designing the experiment

Before we start, we need to figure out three things:

1. What do we mean by one trial?
2. What is the outcome of interest from the trial?
3. How do we simulate one trial?

We then take **three steps** to calculate the desired probability:

1. *Repeat* the simulated trial procedure N times.
2. *Count* M, the number of trials with an outcome that matches the outcome we
   are interested in.
3. Calculate the *proportion*, M/N.
   This is an estimate of the probability in question.

For this problem, our task is made a little easier by the fact that our *trial* (in the resampling sense) is a simulated *trial* (in the legal sense). One trial requires 12 simulated jurors, each labeled by race (White or Black).

The outcome we are interested in is the number of Black jurors.

Now comes the harder part.  How do we simulate one trial?

### One trial

One trial requires 12 jurors, and we are interested only in the race of each juror.

> In our hypothesized world, where selection by race is entirely random, each juror

Reviewer: Maybe?

```suggestion
In Hypothetical County, where selection by race is entirely random, each juror
```

the integers starting at 0 and including 5.
:::

::::

#### Random numbers from 0 through 99

We can now use {{< var np_or_r >}} and the random number function from the last chapter to get 12 random numbers from 0 through 99.

```{python, opts.label="py_ed"}
# Ask NumPy for a random number generator.
rnd = np.random.default_rng()

# Get 12 random numbers from 0 through 99
# (from 0, up to, but not including 100).

> a = rnd.integers(0, 100, size=12)

Reviewer: Should we worry about keyword arguments here?   I mean, explaining them? Or should we defer to a later chapter?

Author: I think the code reads more clearly with them, but they should definitely be explained where we talk about functions.  For now, the reader will just have to infer that it tells the random integer tumbler to tumble out 12 numbers.

```

Notice the argument `5` to the [`np.zeros`]{.python}[`integer`]{.r} function.
This tells the function how many zeros we want in the {{< var array >}}
that the function will return.

The are various useful things we can do with this {{< var array >}} container.
One is to ask how many elements there are in the {{< var array >}} container.

> We can use the [`len`]{.python}[`length`]{.r} function to calculate the number of elements in a[n]{.python} {{< var array >}}:

Reviewer: ```suggestion
We can use the [`len`]{.python}[`length`]{.r} function to calculate the number of elements in {{< var an_array >}}:
```

Of course we could just keep doing this: run the {{< var cell >}} corresponding
to a trial, above, to get a new count, and then store it at the next position
in the `z` {{< var array >}} with:
[`z[1] = count`]{.python}[`z[2] <- count`]{.r},
and so on, but that also looks like a task that the computer is well-suited
for.  And indeed it is, because {{< var lang >}} has an excellent way of
repeating the same set of steps many times — the `for` loop.  You have already seen a
preview of the `for` loop in @sec-resampling-method. Here we dive into them in
more depth.

:::{.callout-box}
### For-loops in {{< var lang >}}

> <!-- Matthew wants to switch this to integers instead of strings; S needs to find out why -->

Reviewer: Because we haven't introduced strings yet ...

Author: I think strings are quite intuitive here; we could mention them earlier if you feel they need explicit introduction.

Reviewer: I think it's fine if we introduce them properly, but strings are a significant thing.  At the moment we just have numbers.  So I think we'd have to work out where we've got the space to put the introduction to strings.

Author: Where do we note these TODOs?

Reviewer: How about `notes/todo.md`?

    print('I am stitching a', item)
```

```{r}
items <- c('shoe', 'saddle', 'jacket')
for (item in items) {
  message('I am stitching a ', item)
}
```

Here, the robot is working its way through three items: a shoe, a saddle, and a jacket.
Every time that a new object arrives in front of it, that object is assigned the label `item`.
The [indented block of code under]{.python}[block of code inside]{.r} the `for` is executed for each one, and in this code block `label` can be accessed.

> You will also notice us using a new function: [`print`]{.python}[`message`]{.r}. This function receives several arguments, and then prints them out to screen.

Reviewer: Why `message` and not `print` for R?   A little research - https://www.geeksforgeeks.org/printing-output-of-an-r-program - suggests that `message` is a bit unusual.

Author: Because `print` requires `print(paste(...))`.

Author: Yes, that's correct, you cannot do `print("Hello", x)` in R. :man_shrugging:

Reviewer: Hmm - yes - ugly - I agree then, that message makes more sense.  But then - I guess we should be using it everywhere.

Author: I am no R expert, so I fully defer to you on what to use in that world.

Reviewer: I think it's OK to stick to message for this - I haven't seen it before, but I see the problem here.   I just mean that we should be consistent.

    print('I am stitching a', package)
```

```{r}
items <- c('shoe', 'saddle', 'jacket')
for (package in items) {
  message('I am stitching a ', package)
}
```

It is also common to operate on numbers in a for loop:

```{python, opts.label="py_ed"}
for i in [0, 1, 2, 3, 4]:
    print('The square of', i, 'is', i * i)
```

```{r}
numbers = c(0, 1, 2, 3, 4)
for (i in numbers) {
  message('The square of ', i, ' is ', i * i)
}
```

When we want to count to a hundred, it becomes laborious to type out the

> numbers. For that, {{< var lang> }} has a shorthand [called `range`]{.python}—very similar to NumPy's `arange`]{.python}:

Reviewer: I think the markup here is a bit messed up? (closing ] without opening, for example)?

We'll also need to consider whether we want to use `range` or `np.arange`.  Berkeley uses `np.arange`, because you don't have to explain:

```
>>> range(10)
range(0, 10)
```

- and that seems like a good idea.  But of course, later on, the students should probably be using `range`...

Author: Leaving that for resolving later.

Finally, we need to count how many of the trials in `z` ended up with all-White juries.  These are the trials with a `z` value of 0.

We can ask this question of *all 100 counts* by asking the question: is the {{< var array >}} `z` equal to 0, like this:

```{python, opts.label="py_ed"}
# Is the value of z equal to 0?

> all_white = (z == 0)

Reviewer: Do we want a previous example on the lines of:

```
>>> range(10)
range(0, 10)
>>> one_z = 3
>>> one_z == 0
False
>>> two_z = np.array([3, 0])
>>> two_z == 0
array([False,  True])
```

Author: Added an example.

# Show the result of the comparison.
all_white
```

```{r, opts.label="r_ed"}
# Is the value of z equal to 0?

> all_white <- (z == 0)

Reviewer: Why the parens?

Author: For newcomers, we've found that masks are easier to understand this way.  Otherwise there can be confusion between operator precedence.  They may read it as: `(all_white <- z) == 0`, which makes no sense.

Reviewer: Who is "we" there?  Scikit-image?

If 'twere me - I'd be asking - why the brackets?   Do they do something?   What would happen if there were no brackets there?

I think the reader won't be very good yet at thinking what the brackets do, and will be tempted to think that they are some sort of magic here, and so introduce confusion.

Author: Readers have brackets from mathematics at early school level, so I don't think it will confuse them nearly as much as the lack of grouping.

Reviewer: Well - but here that's pretty opaque.  For example, they won't be used to 

a = (b + c)

That would look very odd to them - and they would surely wonder - "Why the brackets there?  They don't do anything".  And they already know that parentheses mean calling a function.  So my guess would be pretty hard to the other direction (more confusion than clarity).

Author: I think you're wrong, but I am going to change it anyway.

Author: ```
all_white = z == 0
```

vs

```
all_white = (z == 0)
```

In my mind the second clearly reads better, and that is also the style we adopt inside most libraries AFAIK.  It is also the recommended style by Robert Kern :)

Author: Done.

Reviewer: I think I haven't come across the recommendation - there's probably some source for it somewhere?   And I certainly haven't used it up til now, but I'm sure you've guess that from my question.  But in any case- I think it's a different situation if you know your way around a language.  In that case:

a = (b + c)

won't trip you up, because you're used to reading through brackets.  Although, you saw, I wondered what they were for, and it's hard for me to imagine that beginner readers wouldn't be the same.  I understand the recommendation, and with the explanation, it makes sense, but I think we probably don't want to add an explanation here, and without it, I still think it's just going to be confusing.

Reviewer: Expanding - I think any careful student would ask "Why the brackets?".  And I think we don't want to explain at this point.

Author: As I mentioned, I think the confusion will be around operator precedence; but I've made the change, so let's continue.

Reviewer: Sure - I understood the precedence point.

I think we can deal with that fairly easily with examples like:

```
>>> one_z = 3
>>> one_z == 0
False
>>> two_z = np.array([3, 0])
>>> two_z == 0
array([False,  True])
>>> res = two_z == 0
>>> res
array([False,  True])
```

Author: I think simply showing the result of `np.array([...]) == 3` is sufficient.  A concern I have more widely is that very verbose code is harder to parse and learn from.  Sometimes, better to rely on the students inspecting an output and developing intuition from the result.

Reviewer: Is it possible that this is a difference between the students I have taught and the ones you have taught?  My students are very easily spooked, when they see something they don't understand, and they have very little confidence in their ability to parse code.  That's where my instinct comes from, to really go the long way round at first, in order to pick up speed later.

Author: I guess I use myself as a test case first. I get confused by volume more easily than I do by syntax. But you may well be right that we are used to different audiences.
